packages/liveblocks-core/src/connection.ts

```diff
@@ -412,54 +418,103 @@ function createConnectionStateMachine<T extends BaseAuthResult>(
-           *
-           * When an "open" event happens, we're OK and ready to let the
-           * Promise succeed, and move to the OK state. Before that happens, we
-           * attach the _actual_, more permanent event listeners that will
-           * outlive this state and be used in the OK state mostly.
-           */
-          const socket = delegates.createSocket(ctx.token as T);
-
-          function reject() {
-            socket.removeEventListener("message", onSocketMessage);
-            rej();
+            //
+            // Part 1:
+            // The `error` and `close` event handlers marked (*) are installed
+            // here only temporarily, just to handle this promise-based state.
+            // When those get triggered, we reject this promise.
+            //
+            socket.addEventListener("message", onSocketMessage);
+            socket.addEventListener("error", reject); // (*)
+            socket.addEventListener("close", reject); // (*)
+            socket.addEventListener("open", () => {
+              //
+              // Part 2:
+              // The "open" event just fired, so the server accepted our
+              // attempt to connect. We'll go on and resolve() our promise as
+              // a result.
+              //
+              // However, we cannot safely remove our error/close rejection
+              // handlers _just yet_. There is a small, unlikely-but-possible
+              // edge case: if (and only if) any close/error events are
+              // _already_ queued up in the event queue before this handler is
+              // invoked, then those will fire before our promise will be
+              // resolved.
+              //
+              // Scenario:
+              // - Event queue is empty, listeners are installed
+              // - Two events synchronously get scheduled in the event queue: [<open event>, <close event>]
+              // - The open handler is invoked (= this very callback)
+              // - Event queue now looks like: [<close event>]
+              // - We happily continue and resolve the promise
+              // - Event queue now looks like: [<close event>, <our resolved promise>]
+              // - Close event handler fires, but we already resolved promise! 😣
+              //
+              // This is what I'm calling a "premature" event here, we'll deal with it in part 3.
+              //
+              socket.addEventListener("error", onSocketError);
+              socket.addEventListener("close", onSocketClose);
+              const unsub = () => {
+                socket.removeEventListener("error", reject); // Remove (*)
+                socket.removeEventListener("close", reject); // Remove (*)
+              };
+
+              // Resolve the promise, which will take us to the @ok.* state
+              resolve([socket, unsub]);
+            });
           }
+        );
+
+        return withTimeout(promise, SOCKET_CONNECT_TIMEOUT).then(
```

Comment:
does this need a catch somewhere?

Reply:
It does not! The whole point here is that the promise _should_ fail in all cases but the happy path. When that happens, the machine will [take the failure transition](https://github.com/liveblocks/liveblocks/blob/208f9c0531b3e9e69e76d33ddfc19d414970ae48/packages/liveblocks-core/src/connection.ts#L529-L564) out of that state, and either retries-with-backoff, or (purposefully) stops retrying.

Reply:
That said, I kinda hate how long this `.onEnterAsync()` definition has become. Makes the structure of that call really hard to follow now. The structure is:

```
    .onEnterAsync(
      "@connecting.busy",
      tryConnectingTheSocket,  // a Promise
      ok => { /* transition like this */ },
      failure => { /* transition like this */ },
    )
```

Reply:
in the failure case, what happens to the socket? Is it cleaned up? Are those reject listeners removed?

Reply:
Great question! If the promise succeeds, we retain a permanent reference to the socket, via [this assignment](https://github.com/liveblocks/liveblocks/blob/208f9c0531b3e9e69e76d33ddfc19d414970ae48/packages/liveblocks-core/src/connection.ts#LL524-L524). Otherwise, the socket instance will get garbage collected, so my understanding is that there is no need to remove the listeners.